fpmsyncd/routesync.cpp

```diff
@@ -30,6 +54,466 @@ RouteSync::RouteSync(RedisPipeline *pipeline) :
+            /* This is a multipath route */
+            /* Need to add the code for multipath */
+            int len;
+            struct rtattr *subtb[RTA_MAX + 1];
+            struct rtnexthop *rtnh = (struct rtnexthop *)RTA_DATA(tb[RTA_MULTIPATH]);
+            len = (int)RTA_PAYLOAD(tb[RTA_MULTIPATH]);
+
+            for (;;) 
+            {
+                uint16_t encap = 0;
+                if (len < (int)sizeof(*rtnh)
+                    || rtnh->rtnh_len > len)
+                    break;
+
+                gate = 0;
+                if (rtnh->rtnh_len > sizeof(*rtnh)) 
+                {
+                    memset(subtb, 0, sizeof(subtb));
+
+                    netlink_parse_rtattr(subtb, RTA_MAX, RTNH_DATA(rtnh),
+                                          (int)(rtnh->rtnh_len - sizeof(*rtnh)));
+
+                    if (subtb[RTA_GATEWAY])
+                        gate = RTA_DATA(subtb[RTA_GATEWAY]);
+
+                    if (gate)
+                    {
+                        if (RTA_PAYLOAD(subtb[RTA_GATEWAY]) <= IPV4_MAX_BYTE)
+                        {
+                            memcpy(gateaddr, gate, IPV4_MAX_BYTE);
+                            gw_af = AF_INET;
+                        }
+                        else
+                        {
+                            memcpy(ipv6_address.s6_addr, gate, IPV6_MAX_BYTE);
+                            gw_af = AF_INET6;                    
+                        }
+                    }
+                    
+                    if(gw_af == AF_INET6)
+                    {
+                        if (IN6_IS_ADDR_V4MAPPED(&ipv6_address))
+                        {
+                            memcpy(gateaddr, (ipv6_address.s6_addr+12), IPV4_MAX_BYTE);
+                            gw_af = AF_INET;
+                        }
+                        else
+                        {
+                            SWSS_LOG_NOTICE("IPv6 tunnel nexthop not supported Nexthop:%s encap:%d encap_value:%d",
+                                            inet_ntop(gw_af, ipv6_address.s6_addr, nexthopaddr, MAX_ADDR_SIZE), encap, encap_value);
+                            return false;
+                        }
+                    }
+                    
+                    inet_ntop(gw_af, gateaddr, nexthopaddr, MAX_ADDR_SIZE);
+
+                    if (subtb[RTA_ENCAP_TYPE])
+                        encap = *(uint16_t *)RTA_DATA(subtb[RTA_ENCAP_TYPE]);
+
+                    if (subtb[RTA_ENCAP] && subtb[RTA_ENCAP_TYPE]
+                        && (*(uint16_t *)RTA_DATA(subtb[RTA_ENCAP_TYPE]) == NH_ENCAP_VXLAN))
+                    {
+                        parse_encap(subtb[RTA_ENCAP], encap_value, rmac, vlan);
+                    }
+                    SWSS_LOG_DEBUG("Multipath Nexthop:%s encap:%d encap_value:%d rmac:%s vlan:%d",
+                                    nexthopaddr, encap, encap_value, rmac.c_str(), vlan);
+
+                    if (encap_value == 0 || vlan == 0
```

Comment:
Partial ecmp nexthops are added to route APPDB in this error case. There could be other resolved nhops in the message that are not processed, I think we should process those and add to the DB.

Reply:
We shouldn't program nexthop with partial rewrite into AppDB, but this message could contain valid nexthops after one partial nexthop in the list. If this code returns for one partial nexthop, then valid nexthops are also not added to DB. Is this expected??

Reply:
We don't want to program route even when one partial nexthop information is available. 
If this case exist we need to debug why this incorrect nexthop is send for this prefix(even if one nexthop is not correct), else there will be inconsistency between the control plane and data plane.

Reply:
Makes sense to bypass route addition with partial info. When partial nexthop gets valid encap/mac, does kernel notify existing route RTM_NEWROUTE notification with all nexthop list?? I believe VNID encap and rmac will be always valid as that is part of vtep interface configuration. How about vlan? This could be invalid at the time of route addition.

Reply:
These message are coming directly from FRR not via kernel. In FRR function netlink_route_info_encode encode these message and send to fpm which is received to fpmsyncd. Since these route are directly send by FRR it is better to valid all the information before programming the ASIC.